### PR TITLE
[TA-457] Add cart items to transaction API

### PR DESCRIPTION
**Pull request by steve-gr4vy**
# Description

https://gr4vy.atlassian.net/browse/TA-457

Add cart items to create transaction

Brief: https://docs.google.com/document/d/1puwhGLrl2ujXDgNcxuVggCVqJHpUKmlUw0MRaJ5O7v4/edit?usp=sharing

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own changes
- [x] I have run `yarn test` to make sure my changes pass all linters
- [x] I have pulled the latest changes from the upstream `main` branch

## Contribution guidelines

For contribution guidelines, styleguide, and other helpful information please
see the `CONTRIBUTING.md` file in the root of this project.

### CODE REVIEW

File: reference/resources/cart_item.yaml [hunk 1]
```diff
@@ -0,0 +1,128 @@
+---
+title: Cart Item
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:2]: Should we add something the item's line number? Usually a line number is assigned to an item within a basket. This sometimes help to match products between multiple requests.

**Review by cbetta — APPROVED** [review]

File: reference/resources/cart_item.yaml (hunk 1 above)
Comment (cbetta) [reply reference/resources/cart_item.yaml:2]: After evaluation of PSP APIs this is not required.

**Review by cbetta — APPROVED** [review]

File: reference/resources/cart_item.yaml [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  A cart item that represents a single cart line item for a transaction.
+
+required:
+  - name
+  - quantity
+  - unit_amount
+  - currency
+
+properties:
+  name:
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:14]: @szeying-lee , do you remember which is usually the max length for PSPs?

File: reference/resources/cart_item.yaml [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    type: string
+    description: The name of the cart item.
+    example: GoPro HERO9 Camcorder
+
+  quantity:
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:19]: Should we allow `0`? I assume we shouldn't allow negative values.

File: reference/resources/cart_item.yaml [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    description: |-
+      The amount discounted for this item represented as a monetary amount
+      in the smallest currency unit for the given currency, for example `1299`
+      USD cents represents `$12.99`.
+    example: 0
+    minimum: 0
+    maximum: 99999999
+    nullable: true
+
+  tax_amount:
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:66]: @cbetta 

Should we set a default value for this field (same for `discount_amount`) as zero? Or just mention for these fields that the value may be null or zero depending on the PSP.

File: reference/resources/cart_item.yaml (hunk 3 above)
Comment (cbetta) [reply reference/resources/cart_item.yaml:26]: 👎🏻 Positive integers only.

File: reference/resources/cart_item.yaml (hunk 4 above)
Comment (cbetta) [reply reference/resources/cart_item.yaml:59]: Yeah I think we can set it to 0. Some PSPs will support this field, some wont

File: reference/resources/cart_item.yaml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    example: GoPro HERO9 Camcorder
+    maxLength: 255
+
+  quantity:
+    type: integer
+    description: |-
+      The quantity of this item in the cart. This value cannot be negative
+      or zero.
+    example: 1
+    minimum: 1
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:32]:
```suggestion
    minimum: 1
    maximum: 99999999
```

File: reference/resources/cart_item.yaml [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+  image_url:
+    type: string
+    format: url
+    example: https://example.com/images/items/gopro.png
+    description:
+      The URL for the image of the item.
+    nullable: true
+
+  categories:
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:107]: We should probably set a sensible limit on this one.

@cbetta , do you have any preferences? Can't remember if we have something similar to this in the API, is 10 or 15 a good limit?

File: reference/resources/cart_item.yaml [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    type: integer
+    description: |-
+      The amount for an individual item represented as a monetary amount
+      in the smallest currency unit for the given currency, for example
+      `1299` USD cents represents `$12.99`.
+    example: 37999
+    minimum: 0
+    maximum: 99999999
+
+  currency:
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:45]: Do we need this? Can we assume that currency is that same as the currency for the transaction?

CC: @cbetta @steve-gr4vy

File: reference/resources/cart_item.yaml [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+  tax_amount:
+    type: integer
+    description: |-
+      The tax amount for this item represented as a monetary amount
+      in the smallest currency unit for the given currency, for example `1299`
+      USD cents represents `$12.99`.
+
+      Important: this amount is for the total of the cart item and not for
+      the individual item.
```
Comment (szeying-lee) [inline reference/resources/cart_item.yaml:77]:
```suggestion
      Important: this amount is for the total of the cart item and not for
      an individual item. For example, if the quantity is 5, this value
      should be the total tax amount for 5 cart items.
```

File: reference/resources/cart_item.yaml [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+  discount_amount:
+    type: integer
+    description: |-
+      The amount discounted for this item represented as a monetary amount
+      in the smallest currency unit for the given currency, for example `1299`
+      USD cents represents `$12.99`.
+
+      Important: this amount is for the total of the cart item and not for
+      the individual item.
```
Comment (szeying-lee) [inline reference/resources/cart_item.yaml:62]:
```suggestion
      Important: this amount is for the total of the cart item and not for
      an individual item. For example, if the quantity is 5, this value
      should be the total discount amount for 5 cart items.
```

File: reference/resources/cart_item.yaml [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+  If total due to be paid for the item is required by the payment service
+  provider, sometimes called "total amount", formula below will usually
+  be used to calculate this amount:
+
+  `(unit_amount * quantity) - discount_amount + tax_amount`
+
+  It's highly recommended that the total price to pay for all items
+  should match the transaction's amount to reduce the risk of the
+  transaction being declined by the payment service provider.
```
Comment (szeying-lee) [inline reference/resources/cart_item.yaml:18]:
```suggestion
  If the total due to be paid for the item is required by the payment service
  provider, generally referred to as the "total amount", the formula below
  will usually be used to calculate this amount:

  `(unit_amount * quantity) - discount_amount + tax_amount`

  It's highly recommended that the total amount to pay for all items
  should match the transaction's amount to reduce the risk of the
  transaction being declined by the payment service provider.
```

File: reference/resources/cart_item.yaml (hunk 9 above)
Comment (szeying-lee) [inline reference/resources/cart_item.yaml:62]:
```suggestion
      Important: this amount is for the total of the cart item and not for
      an individual item. For example, if the quantity is 5, this value
      should be the total discount amount for 5 of the cart item.
```

File: reference/resources/cart_item.yaml (hunk 8 above)
Comment (szeying-lee) [inline reference/resources/cart_item.yaml:77]:
```suggestion
      Important: this amount is for the total of the cart item and not for
      an individual item. For example, if the quantity is 5, this value
      should be the total tax amount for 5 of the cart item.
```

File: reference/resources/cart_item.yaml [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    description:
+      The URL for the image of the item.
+    nullable: true
+    maxLength: 2083
+
+  categories:
+    type: array
+    description: |-
+      A list of strings containing product categories for the item.
+      Max length: 50.
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:125]: I added this as I couldn't see the `maxLength` limit in the documentation when testing it locally:
image

File: reference/resources/cart_item.yaml [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    description:
+      The URL for the image of the item.
+    nullable: true
+    maxLength: 2083
+
+  categories:
+    type: array
+    description: |-
+      A list of strings containing product categories for the item.
+      Max length per item: 50.
```
Comment (ipeluffo) [inline reference/resources/cart_item.yaml:125]: I added this as I couldn't see the `maxLength` limit in the documentation when testing it locally:
image

**Comment by ipeluffo** [comment]
@cbetta @gryevns @steve-gr4vy , sorry to ask this now but I just realised it while doing some tests.

Should we consider adding a `description` field?

**Comment by gryevns** [comment]
> @cbetta @gryevns @steve-gr4vy , sorry to ask this now but I just realised it while doing some tests.
> 
> Should we consider adding a `description` field?

You can find the spike research that Sze completed [here](https://gr4vy.atlassian.net/wiki/spaces/GB/pages/207028271/TA-502+Research+cart+item+support) - no

> @cbetta @gryevns @steve-gr4vy , sorry to ask this now but I just realised it while doing some tests.
> 
> Should we consider adding a `description` field?

I think `name` & `description` are interchangeable, [here](https://gr4vy.atlassian.net/wiki/spaces/GB/pages/207028271/TA-502+Research+cart+item+support) is the spike research doc.

**Review by gryevns — APPROVED** [review]

**Pull request merged**